### ISSUE TITLE
literal not inferred for staticmethod resulting in partially unknown

### ISSUE THREAD

**Issue by leonardgerardatomicmachinescom** [microsoft/pyright#10555]
**Describe the bug**

I would assume the literal of type `L` below to be resolved by pyright when calling the static method to create `c`, but it does not.

```python
import typing

L = typing.TypeVar("L", bound=typing.Literal[1, 2])

class C(typing.Generic[L]):
    def __init__(self, l: L):
        self.l = l

    @staticmethod
    def m(l: L) -> C[L]:
        return C(l)

c = C.m(1)
```

Errors:

```
error: Type of "c" is partially unknown
    Type of "c" is "C[Unknown]" (reportUnknownVariableType)
error: Type of "m" is partially unknown
    Type of "m" is "(l: Unknown) -> C[Unknown]" (reportUnknownMemberType)
```

Note:

Wrapping the staticmethod in a function allows pyright to not complain:
```
def f(l: L) -> C[L]:
    return C[L](l)

c = f(1)
```

**Comment by erictraut** [comment]
Pyright is working correctly here. The expression `C.m(1)` is equivalent to `C[Any].m(1)`. There is no default type argument value defined for the type parameter `L`, so it gets the default `Any` (or `Unknown`, which is pyright's name for an implicit `Any`). 

You can achieve what you want by using a method-scoped type variable rather than the class-scoped type variable `L`.

```python
class C[L: Literal[1, 2]]:
    def __init__(self, l: L):
        self.l = l

    @staticmethod
    def m[X: Literal[1, 2]](l: X) -> "C[X]":
        return C(l)
```

**Comment by leonardgerardatomicmachinescom** [comment]
As a user, it is surprising to not have type inference work here. When I do not specify the type parameters for C, I expect the rest of the expression to constrain it.

The error messages were not helpful.

In particular, the error message indicates `C[Unknown]` and not `C[Any]`. The latter would have led me to try to scope at the method.

**Comment by erictraut** [comment]
Type inference isn't involved in this case. The values of class-scoped type parameters are inferred only for constructor calls, and this isn't a constructor. The method's type is partially specialized when it is bound to `C` in the expression `C.m`. In other words, the type of the expression `C.m` is `(l: Unknown) -> C[Unknown]`. By the time the call expression is evaluated by the type checker, there are no type variables left to solve.

This behavior might make more sense if you read [PEP 696](https://peps.python.org/pep-0696/), which added support for type parameter defaults. If no default is supplied, then the type argument becomes `Unknown`. If you'd like to specify a default type parameter value, you can do so like this:

```python
from typing import Generic, Literal, TypeVar

L = TypeVar("L", bound=Literal[1, 2], default=Literal[1])

class C(Generic[L]):
    def __init__(self, l: L):
        self.l = l

    @staticmethod
    def m(l: L) -> C[L]:
        return C(l)

reveal_type(C.m) # (l: Literal[1]) -> C[Literal[1]]

c = C.m(1)
reveal_type(c) # C[Literal[1]]

C.m(2) # Type error

C[Literal[2]](2) # OK
```

If you're using Python 3.14 beta, you can use the new syntax for default type parameter values:

```python
class C[L: Literal[1, 2] = Literal[1]]:
    ...
```

**Comment by leonardgerardatomicmachinescom** [comment]
> The values of class-scoped type parameters are inferred only for constructor calls, and this isn't a constructor.

This is the key point. Is this a pyright specific decision?


(Is there a technical reason to differentiate between constructors and classmethod or staticmethod?)


PS: I am stuck with python 3.11

**Comment by erictraut** [comment]
> Is this a pyright specific decision?

The spec is admittedly incomplete when it comes to method binding behaviors, but pyright's behavior is strongly implied by PEP 696. Without this interpretation, PEP 696 would be inconsistent.

```python
class A[T = str]: # Default is str
    @classmethod
    def meth(cls) -> T: ...

class B[T]: # Default is Unknown
    @classmethod
    def meth(cls) -> T: ...

reveal_type(A) # type[A[str]]
reveal_type(B) # type[B[Unknown]]

reveal_type(A.meth) # () -> str
reveal_type(B.meth) # () -> Unknown
```

If you want the return type of a bound static or class method to be inferred based on arguments passed to the call, there's a way to do that — by using a method-scoped type variable as I showed above. If you use a class-scoped type variable, that has a different meaning and supports different use cases. I don't think it makes sense to treat both cases the same.

**Comment by leonardgerardatomicmachinescom** [comment]
One reason I would expect (at least classmethod) to behave like constructors is that classmethods are very often used as factory methods, effectively playing the role of constructors:

```python
from typing import Generic, Literal, TypeVar, Self

L = TypeVar("L", bound=Literal[1, 2])

class C(Generic[L]):
    def __init__(self, l: L):
        self.l = l

    @classmethod
    def from_l(cls, l: L) -> Self:
        return cls(l)

c = C.from_l(1) # typical factory approach
```


Why would this contradict PEP 696 isn't clear to me yet.

**Comment by leonardgerardatomicmachinescom** [comment]
PS: I changed the title, since it is not specific to literals, even though they make the example easier.

PS 2: It does feel in the same vein as my previous ["bug" report](https://github.com/microsoft/pyright/issues/9870)
